include/wil/registry_helpers.h

```diff
@@ -0,0 +1,947 @@
+//
+//*********************************************************
+#ifndef __WIL_REGISTRY_HELPERS_INCLUDED
+#define __WIL_REGISTRY_HELPERS_INCLUDED
+
+#if defined(_STRING_) || defined (_VECTOR_) || (defined (__cpp_lib_optional) && defined (_OPTIONAL_))
+#include <functional>
+#include <iterator>
+#endif
+
+#include <Windows.h>
+#include "resource.h"
+
+#ifdef _KERNEL_MODE
+#error This header is not supported in kernel-mode.
+#endif
+
+#ifndef __WIL_WINREG_
+#error This is required for ::wil::unique_hkey support
+#endif
+
+namespace wil
+{
+    namespace reg
+    {
+        enum class key_access
+        {
+            read,
+            readwrite,
+        };
+
+        namespace reg_view_details
+        {
+            constexpr DWORD get_value_flags_from_value_type(DWORD type) WI_NOEXCEPT
+            {
+                switch (type)
+                {
+                case REG_DWORD:
+                    return RRF_RT_REG_DWORD;
+                case REG_QWORD:
+                    return RRF_RT_REG_QWORD;
+                case REG_SZ:
+                    return RRF_RT_REG_SZ | RRF_RT_REG_EXPAND_SZ | RRF_NOEXPAND;
+                case REG_EXPAND_SZ:
+                    return RRF_RT_REG_SZ | RRF_RT_REG_EXPAND_SZ;
+                case REG_MULTI_SZ:
+                    return RRF_RT_REG_MULTI_SZ;
+                case REG_BINARY:
+                    return RRF_RT_REG_BINARY;
+                    // the caller can directly specify their own flags if they need to
+                default:
+                    return type;
+                }
+            }
+
+            constexpr DWORD get_access_flags(key_access access) WI_NOEXCEPT
+            {
+                switch (access)
+                {
+                case key_access::read:
+                    return KEY_READ;
+                case key_access::readwrite:
+                    return KEY_ALL_ACCESS;
+                }
+                FAIL_FAST();
+            }
+
+#if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+            template <class InputIt>
+            ::std::vector<wchar_t> get_multistring_from_wstrings(const InputIt& first, const InputIt& last)
+            {
+                ::std::vector<wchar_t> multistring;
+
+                if (first == last)
+                {
+                    multistring.push_back(L'\0');
+                    multistring.push_back(L'\0');
+                    return multistring;
+                }
+
+                for (const auto& wstr : ::wil::make_range(first, last))
+                {
+                    if (!wstr.empty())
+                    {
+                        // back_inserter returns an iterator to enable ::std::copy to write into 'wstr'
+                        // from begin() to end() at the back_inserter iterator, growing the vector as necessary
+                        ::std::copy(::std::begin(wstr), ::std::end(wstr), ::std::back_inserter(multistring));

```

Comment:
Does this devolve into `multistring.insert(multistring.end(), begin(wstr), end(wstr))` ? Or does it end up calling "push_back" repeatedly?

Reply:
a quick read of the STL suggests it calls push_back repeatedly. I can't find something that's more optimized to directly take the begin & end iterators and directly resizing to write them in.

Reply:
which means I can greatly simplify (and optimize) this based off your suggestion :)

and all our crazy multi-string tests continue to pass, even with verifier enabled.